onc-healthit/lantern-back-end
# Implement delete fhir endpt logic

This PR implements the logic for deleting fhir endpoints based on update times following M2/K1 as described here:
https://mitre.sharepoint.com/sites/ProjectLantern/_layouts/15/Doc.aspx?sourcedoc={bc496df4-9591-43a5-8860-8828979cb683}&action=edit&wd=target%28Development%20Plan.one%7C273069de-a04d-41e2-9f83-5cb18ce5d5e5%2FMaintaining%20list%20of%20current%20endpoints%7C56dac67e-e104-3741-ba05-14571baeed01%2F%29

To test:
This is easier to test if use just the CareEvolution list source and comment out the others. There will still be some endpoints from NPPES.  
1. run `make populatedb`
2. `cd endpointmanager/cmd` and run `go run sendendpoints/main.go`
3. In same directory but different terminal run `go run capabilityreceiver/main.go`
4. Check the database `docker exec -it lantern-back-end_pg_prometheus_1 psql -U lantern`
and verify that fhir_endpoints and fhir_endpoints_info have entries in them. Note the updated_at time of the endpoints and the number of endpoints in each table. Also check endpoints that are in the table multiple times from different list source.
`select * from fhir_endpoints where url in (select url from fhir_endpoints group by url having count(*) > 1) order by url;`
5. edit the populatedb.sh and replace the line
`go run main.go ../../../networkstatsquerier/resources/EndpointSources.json CareEvolution`
with:
`go run main.go ../../../networkstatsquerier/resources/EndpointSources_1.json CareEvolution`
6. run `make populatedb.sh`
7. Check the database and verify that there are 10 less endpoints in fhir_endpoints. fhir_endpoints_info table will have 6 less endpoints because of the 4 of the 10 endpoints that were removed in EndpointSources_1.json still exist from the NPPES list source. Check the fhir_endpoints_info_history table to see the deleted endpoints. 
`select operation, url, updated_at from fhir_endpoints_info_history where operation='D';`

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: https://oncprojectracking.healthit.gov/support/browse/LANTERN-198
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code accomplishes the tasks purpose.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge
is complete.

Create a new branch `<branch name>_gomodupdate`. Run `make update_mods` to update the `go.mod`
and `go.sum` files to point to master. Create a PR. Require one sanity check reviewer.
